### PR TITLE
Fixes #42: Move hashtags to redux store

### LINKED ISSUE #42: Show related hashtags on bottom of search result page
On the bottom of search page results we often find "related searches". Please implement a similar thing "related hashtags".

screenshot from 2016-10-17 17 18 12

### PR DESCRIPTION
**Pull request by uttu357**
- Compute tags from API response
- Add tags array to redux state of API response

### CODE REVIEW

**Comment by hemantjadon** [comment]
@uttu357 Few suggestions here:

Move the tag computation logic out of the `effects` into the reducer function which takes it directly into the state.

What i mean there should be no logic in `effects` related to state.

What moves into the state is reducers task not the effects. 

Pass the reducer the complete ApiResponse object and let reducer handle it in SEARCH_COMPLETE_SUCCESS. That will be clearer. and concise

File: src/app/effects/api-search.effects.ts [hunk 1]
```diff
@@ -49,6 +49,15 @@ export class ApiSearchEffects {
 																					if (query.location === ReloactionAfterQuery.RELOCATE) {
 																						this.location.go(`/search?query=${query.queryString}`);
 																					}
+																					let tagShards = [].concat(...response.statuses.map((a) => (a.hashtags)));
```
Comment (hemantjadon) [inline src/app/effects/api-search.effects.ts]: This goes out of here to the reducer.

**Comment by uttu357** [comment]
@hemantjadon Thanks for your suggestions! Have made the changes. Please review them.

**Comment by hemantjadon** [comment]
@uttu357 This seems much better, just one more thing which I guess would be appropriate is removing the `ApiResponseTags` from `models/api-response.ts`.

What I think is model should only contain the data which is coming from the server or a backend service not the one which is generated by us. I don't feel the need for it.

Tags array can be a simple key (of  type `[]`) in state. No need of that in the models.

If we want type checking something like this can be done

```typescript
export interface State {
    . . . 
    hash: Array<{ tag: string, count: number }>;
}
```

**Comment by uttu357** [comment]
Although not related to this PR but what is the need of app/shared/classes if models are present?

**Comment by hemantjadon** [comment]
@uttu357 They were used earlier and are not used now Please remove those files. :)

**Comment by uttu357** [comment]
Should i remove them in this PR only?

**Comment by hemantjadon** [comment]
You can do it here.
Also you will need to squash the commits so that we can merge easily.

**Comment by uttu357** [comment]
@hemantjadon done :)

File: src/app/actions/api.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import { Action } from '@ngrx/store';
-import { ApiResponse } from '../models/api-response';
+import { ApiFetchResponse } from '../models/api-response';
```
Comment (hemantjadon) [inline src/app/actions/api.ts]: Can we have ApiResponse instead of ApiFetchResponse?

File: src/app/actions/api.ts [hunk 3]
```diff
@@ -34,7 +34,7 @@ export class SearchAction implements Action {
 export class SearchCompleteSuccessAction implements Action {
 	type = ActionTypes.SEARCH_COMPLETE_SUCCESS;
 
-	constructor(public payload: ApiResponse) { }
+	constructor(public payload: ApiFetchResponse) { }
```
Comment (hemantjadon) [inline src/app/actions/api.ts]: And here also

File: src/app/models/api-response.ts [hunk 4]
```diff
@@ -62,7 +62,13 @@ export interface ApiResponseResult {
 	without_luh_len: number;
 }
 
+export interface ApiFetchResponse {
```
Comment (hemantjadon) [inline src/app/models/api-response.ts]: We can drop this also I guess>

File: src/app/models/api-response.ts [hunk 5]
```diff
@@ -62,7 +62,13 @@ export interface ApiResponseResult {
...
+	search_metadata: ApiResponseMetadata;
+	statuses: ApiResponseResult[];
+}
+
 export interface ApiResponse {
 	search_metadata: ApiResponseMetadata;
 	statuses: ApiResponseResult[];
+	tags: Array<{ tag: string, count: number }>;
```
Comment (hemantjadon) [inline src/app/models/api-response.ts]: We can just remove this attribute from here? It will work fine.

**Comment by uttu357** [comment]
@hemantjadon made the changes.. :)

**Comment by hemantjadon** [comment]
@uttu357 Nice work there 👍

**Pull request merged**